src/lb_data.c

```diff
@@ -1192,6 +1405,44 @@ int lb_halo_create(const lb_t * lb, lb_halo_t * h, lb_halo_enum_t scheme) {
     h->request[ireq] = MPI_REQUEST_NULL;
   }
 
+
+  /* Device */
+
+  int ndevice;
+  tdpGetDeviceCount(&ndevice);
+  tdpStreamCreate(&h->stream);
+
+  if (ndevice == 0) {
+    h->target = h;
+  }
+  else {
+    tdpAssert( tdpMalloc((void **) &h->target, sizeof(lb_halo_t)) );
+    tdpAssert( tdpMemcpy(h->target, h, sizeof(lb_halo_t),
+			 tdpMemcpyHostToDevice) );
+
+    for (int p = 0; p < h->map.nvel; p++) {         
+      int scount = send_count[p]*lb_halo_size(h->slim[p]);  
+      int rcount = recv_count[p]*lb_halo_size(h->rlim[p]);
+      tdpAssert( tdpMalloc((void**) &h->send_d[p], scount * sizeof(double)) );
+      tdpAssert( tdpMalloc((void**) &h->recv_d[p], rcount * sizeof(double)) );
+    }
+    /* Slightly tricksy. Could use send_d and recv_d on target copy ...*/
+    tdpAssert( tdpMemcpy(h->target->send, h->send_d, 27*sizeof(double *),     
+			 tdpMemcpyHostToDevice) );
+    tdpAssert( tdpMemcpy(h->target->recv, h->recv_d, 27*sizeof(double *),
+			 tdpMemcpyHostToDevice) );
+
```

Comment:
I might move the `halo_initialise_device_model()` here, 1) to guard against `ndevice == 0` and 2) to ensure the halo structure is fully initialised before creating the graph.